iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp

```diff
@@ -674,6 +674,81 @@ class ScatterRank0Value : public OpRewritePattern<mhlo::ScatterOp> {
   }
 };
 
+// Mhlo of non-finite values (e.g. NaN, inf) and 0.0 produce 0.0 for XLA. For
+// linalg we need to conver these to select operations.
```

Comment:
This looks fragile to me. Is it a defined behavior in XLA?

Does mhlo.mul op have the same semantic? If so, we may want to fix that in mhlo.mul op lowering. If not, can this be fixed when lowering things into mhlo? E.g., let's don't generate convert(i1->f32) -> mul pattern.

Reply:
@jpienaar should probably chime in on this. It is defined behavior where `0.0 * x -> 0.0` even for `inf` and `NaN`. And we can't correct at a much higher level as this problem is encountered for a JAX model, which directly generate `mhlo`.

The only alternative solution is to handle all `mhlo.mul` operations to perform similar bitwise / select operations for all `mhlo.mul` lowerings. Not sure what our best path is for codegen as the behavior is abnormal.

Reply:
XLA has this pattern https://github.com/tensorflow/tensorflow/blob/294fc72ed56037ec334fa4646204da7d287c6e50/tensorflow/compiler/xla/service/algebraic_simplifier.cc#L2689, so I think this is effectively part of HLO "executable spec". This makes sense to me if rhs was restricted to integers (as higher up in the file is noted). I think this _may_ be something that could get changed. We may need this as a short term thing (and yes its fragile) and then work with folks to address it.

Reply:
Thanks for the pointer! This explains why I only see convert and broadcast check in `IsFromBool`.

It may be possible that the value is from other operation (e.g., constant op). Would be good if we can add a comment and reference it.